### PR TITLE
Integrate erc transfer data for NFTS

### PR DESCRIPTION
**Pull request by masquot**
## Notes for reviewing and merging

This is a major update based on input of of the community (previous PR https://github.com/duneanalytics/abstractions/pull/414):
- integration of data for bundle trades enables more sophisticated analysis
- integration of corresponding erc721, erc1155 or erc20 transfer data for every trade

The table schema has been modified and extended to accomodate extra data points. Therefore, the existing table will need to be deleted, as well as the existing `insert_xxx` functions and cron jobs.

Thank you very much

### CODE REVIEW

**Review by va3093 — CHANGES_REQUESTED** [review]
@masquot Amazing work. This seems like an incredible amount of work. The main thing I would like to add here is a lot more documentation. I am anticipating a nightmare situation where something breaks here and you are on holiday or something and I need to try and fix it. I would need a good bit more explanation throughout to help understand what is happening. Perhaps we should have a call so you can talk me through this and we can see where would be good to add some clarification.

File: ethereum/nft/trades/insert_cryptopunks.sql [hunk 1]
```diff
@@ -3,54 +3,129 @@ LANGUAGE plpgsql AS $function$
...
+WITH punks_union AS (
+SELECT
+    "evt_tx_hash",
+    "evt_block_time",
+    "evt_block_number",
+    "evt_index",
+    'erc20' AS erc_type,
+    CAST("punkIndex" AS text) AS "tokenId",
+    "fromAddress" AS "from",
+    CASE WHEN "toAddress" = '\x0000000000000000000000000000000000000000'
```
Comment (va3093) [inline ethereum/nft/trades/insert_cryptopunks.sql:15]: Specify why you are special casing this address

File: ethereum/nft/trades/insert_cryptopunks.sql [hunk 2]
```diff
@@ -3,54 +3,129 @@ LANGUAGE plpgsql AS $function$
...
+         THEN (SELECT "to"
+               FROM erc20."ERC20_evt_Transfer" e
+               WHERE e."evt_tx_hash" = p."evt_tx_hash" AND e."from" = p."fromAddress" ANd e.contract_address = '\xb47e3cd837ddf8e4c57f05d70ab865de6e193bbb')
```
Comment (va3093) [inline ethereum/nft/trades/insert_cryptopunks.sql:18]: Specify why you are special casing this address

File: ethereum/nft/trades/insert_cryptopunks.sql [hunk 3]
```diff
@@ -3,54 +3,129 @@ LANGUAGE plpgsql AS $function$
...
+    CASE WHEN "value" = 0
+        THEN COALESCE(
+        (
+         SELECT value
+         FROM cryptopunks."CryptoPunksMarket_evt_PunkBidEntered" b
+         WHERE b."punkIndex" = p."punkIndex"
+             AND b."fromAddress" = (
+                SELECT "to"
+                FROM erc20."ERC20_evt_Transfer" e
+                WHERE e."evt_tx_hash" = p."evt_tx_hash" AND e."from" = p."fromAddress" AND e.contract_address = '\xb47e3cd837ddf8e4c57f05d70ab865de6e193bbb'
```
Comment (va3093) [inline ethereum/nft/trades/insert_cryptopunks.sql:31]: Specify why you are special casing this address

File: ethereum/nft/trades/README.md [hunk 4]
```diff
@@ -6,7 +6,6 @@ The goal is to enable a powerful new wave of Dune dashboards on NFTs, and inspir
 
 Known limitations:
 - OpenSea and Rarible support non-ETH trades. Some of the tokens used for these transactions are not in Dune's `prices.usd`. In these cases, the `usd_amount` is `NULL`. At the time of writing (July 2021), for OpenSea and Rarible, the percentage of rows with missing USD values is 2% or less.
-- Some column data for less common trades such as bundle sales or multiple transfers in one transaction will be incorrect e.g. the column `nft_token_id`. On the tests we did, the `buyer` and `amount` fields were correct.
 - In some cases, there are rounding errors on the `original_amount` fields.
```
Comment (va3093) [inline ethereum/nft/trades/README.md:9]: Give an overview here of why you interact with erc721 + erc1155 so much.

File: ethereum/nft/trades/insert_cryptopunks.sql [hunk 5]
```diff
@@ -3,54 +3,129 @@ LANGUAGE plpgsql AS $function$
...
         'ETH' AS original_currency,
         '\x0000000000000000000000000000000000000000'::bytea AS original_currency_contract,
         '\xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2'::bytea AS currency_contract,
-        trades.contract_address AS nft_contract_address,
-        trades.contract_address AS exchange_contract_address,
+        '\xb47e3cd837ddf8e4c57f05d70ab865de6e193bbb'::bytea AS nft_contract_address,
+        '\xb47e3cd837ddf8e4c57f05d70ab865de6e193bbb'::bytea AS exchange_contract_address,
         trades.evt_tx_hash AS tx_hash,
         trades.evt_block_number AS block_number,
+        trades.token_id_array AS nft_token_ids_array, -- new
```
Comment (va3093) [inline ethereum/nft/trades/insert_cryptopunks.sql:123]: what are these new comments?

File: ethereum/nft/trades/insert_foundation.sql [hunk 6]
```diff
@@ -3,40 +3,94 @@ LANGUAGE plpgsql AS $function$
...
+    CAST(erc721."tokenId" AS TEXT) AS "tokenId",
+    erc721."from",
+    erc721."to",
+    erc721.contract_address,
+    NULL::numeric AS value
+FROM erc721."ERC721_evt_Transfer" erc721
+INNER JOIN foundation."market_evt_ReserveAuctionFinalized" f ON erc721.evt_tx_hash = f.evt_tx_hash
+WHERE erc721.evt_block_time >= start_ts
+AND erc721.evt_block_time < end_ts
+AND erc721."from" <> '\x0000000000000000000000000000000000000000'
```
Comment (va3093) [inline ethereum/nft/trades/insert_foundation.sql:19]: Specify why you are special casing this address

File: ethereum/nft/trades/insert_foundation.sql [hunk 7]
```diff
@@ -3,40 +3,94 @@ LANGUAGE plpgsql AS $function$
...
+    CAST(erc1155.id AS TEXT) AS "tokenId",
+    erc1155."from",
+    erc1155."to",
+    erc1155.contract_address,
+    erc1155.value
+FROM erc1155."ERC1155_evt_TransferSingle" erc1155
+INNER JOIN foundation."market_evt_ReserveAuctionFinalized" f ON erc1155.evt_tx_hash = f.evt_tx_hash
+WHERE erc1155.evt_block_time >= start_ts
+AND erc1155.evt_block_time < end_ts
+AND erc1155."from" <> '\x0000000000000000000000000000000000000000'
```
Comment (va3093) [inline ethereum/nft/trades/insert_foundation.sql:33]: Specify why you are special casing this address

File: ethereum/nft/trades/insert_foundation.sql [hunk 8]
```diff
@@ -3,40 +3,94 @@ LANGUAGE plpgsql AS $function$
...
+        trace_address,
+        evt_index,
+        trade_id
     )
 
     SELECT
         trades.evt_block_time AS block_time,
         tokens.name AS nft_project_name,
-        CAST(created."tokenId" AS TEXT) AS nft_token_id,
+        CASE WHEN erc.no_of_transfers > 1 THEN NULL ELSE CAST(created."tokenId" AS TEXT) END AS nft_token_id, -- modified
```
Comment (va3093) [inline ethereum/nft/trades/insert_foundation.sql:88]: maybe explain why this condition exists? Ie. why are you making this check.

Comment (va3093) [reply ethereum/nft/trades/insert_foundation.sql:88]: same below.

File: ethereum/nft/trades/insert_rarible.sql [hunk 9]
```diff
@@ -154,86 +155,136 @@ WITH all_data AS (
...
+    CAST(erc721."tokenId" AS TEXT) AS "tokenId",
+    erc721."from",
+    erc721."to",
+    erc721.contract_address,
+    NULL::numeric AS value
+FROM erc721."ERC721_evt_Transfer" erc721
+INNER JOIN rarible_trades d ON erc721.evt_tx_hash = d.evt_tx_hash
+WHERE erc721.evt_block_time >= start_ts
+AND erc721.evt_block_time < end_ts
+AND erc721."from" <> '\x0000000000000000000000000000000000000000'
```
Comment (va3093) [inline ethereum/nft/trades/insert_rarible.sql:249]: Same as above. "Ignore mint txs " etc...

File: ethereum/nft/trades/insert_rarible.sql [hunk 10]
```diff
@@ -248,24 +299,32 @@ rows AS (
...
 	trace_address,
 	evt_index,
 	trade_id
     )
-
     SELECT
         trades.evt_block_time AS block_time,
         tokens.name AS nft_project_name,
-        trades.nft_token_id,
+        CASE WHEN erc.no_of_transfers > 1 THEN NULL ELSE trades.nft_token_id END AS nft_token_id, -- modified
```
Comment (va3093) [inline ethereum/nft/trades/insert_rarible.sql:317]: Maybe just drop a note explaining why this check exists.

Comment (va3093) [reply ethereum/nft/trades/insert_rarible.sql:317]: same below.

File: ethereum/nft/trades/insert_superrare.sql [hunk 11]
```diff
@@ -137,92 +151,162 @@ UNION ALL
...
+    CAST(erc721."tokenId" AS TEXT) AS "tokenId",
+    erc721."from",
+    erc721."to",
+    erc721.contract_address,
+    NULL::numeric AS value
+FROM erc721."ERC721_evt_Transfer" erc721
+INNER JOIN superrare_trades ON superrare_trades.tx_hash = erc721.evt_tx_hash
+WHERE erc721.evt_block_time >= start_ts
+AND erc721.evt_block_time < end_ts
+AND erc721."from" <> '\x0000000000000000000000000000000000000000'
```
Comment (va3093) [inline ethereum/nft/trades/insert_superrare.sql:167]: Same as previously mentioned and for all below.

Comment (va3093) [reply ethereum/nft/trades/insert_superrare.sql:167]: i.e Specify why you are special casing this address

File: ethereum/nft/trades/trades.sql [hunk 12]
```diff
@@ -18,6 +21,12 @@ CREATE TABLE nft.trades (
     exchange_contract_address bytea NOT NULL,
     tx_hash bytea NOT NULL,
     block_number integer,
+    nft_token_ids_array text[],
```
Comment (va3093) [inline ethereum/nft/trades/trades.sql:24]: Maybe just briefly explain why these arrays exist. I.e "Sometimes many trades occur in a given trade" etc.

**Review by va3093 — APPROVED** [review]
Just dropped some notes on where I think some comments would have been useful for me. Apologies if I am not consistent in my comments, but hopefully you can see the type of things that might be useful for readers. Whenever there is special casing or some conditional logic it can be useful to have some comments if the actual code is not intuitive.

But great work. I am gonna approve, as the actual code looks 👌

File: ethereum/nft/trades/trades.sql (hunk 12 above)
Comment (masquot) [reply ethereum/nft/trades/trades.sql:24]: Thank you so much @va3093. Addressed your comments in a new commit.

**Review by va3093 — APPROVED** [review]
🤩 . Comments are super helpful. Thanks @masquot

**Pull request merged**